osu.Game/Overlays/Profile/Sections/Beatmaps/PaginatedBeatmapContainer.cs

```diff
@@ -51,7 +70,13 @@ protected override void ShowMore()
                     if (!s.OnlineBeatmapSetID.HasValue)
                         continue;
 
-                    var panel = new DirectGridPanel(s.ToBeatmapSet(Rulesets));
+                    bool beatmapExists = manager.QueryBeatmapSet(b => b.OnlineBeatmapSetID == s.OnlineBeatmapSetID) != null;

```

Comment:
this same logic is basically copied+pasted from `DirectOverlay`. We can't just keep implementing this in each place beatmaps are displays (it'll get very ugly).

We may want to consider moving this logic to inside the panel itself.

Reply:
`DirectOverlay`'s application of this logic is to remove/not add the panel itself. I figured having this external to panel is better because then direct can have the ability to decide whether a panel should be removed or just download-disabled. This would allow something like "show already downloaded beatmaps" filter.

Reply:
It's probably a good idea at this point to check out `BeatmapCarousel` and follow similar event logic to how it's being done there (I think it worked quite well). Just hoping that at the end of the day we can somehow combine the `BeatmapSetOverlay` (paginated container) case and the `DirectOverlay` into a single implementation of panel visibility logic.